phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java

```diff
@@ -490,15 +493,37 @@ private static String toString(List<byte[]> gps) {
         }
         List<List<Scan>> parallelScans = Lists.newArrayListWithExpectedSize(stopIndex - regionIndex + 1);
         
-        byte[] currentKey = startKey;
-        int guideIndex = currentKey.length == 0 ? 0 : getIndexContainingInclusive(gps, currentKey);
-        int gpsSize = gps.size();
+        ImmutableBytesWritable currentKey = new ImmutableBytesWritable(startKey, 0, startKey.length);
+        
+        int gpsSize = gps.getGuidePostsCount();
         int estGuidepostsPerRegion = gpsSize == 0 ? 1 : gpsSize / regionLocations.size() + 1;
         int keyOffset = 0;
+        ImmutableBytesWritable currentGuidePost = ByteUtil.EMPTY_IMMUTABLE_BYTE_ARRAY;
         List<Scan> scans = Lists.newArrayListWithExpectedSize(estGuidepostsPerRegion);
+        ImmutableBytesWritable guidePosts = gps.getGuidePosts();
+        ByteArrayInputStream stream = null;
+        DataInput input = null;
+        PrefixByteDecoder decoder = null;
+        int guideIndex = 0;
+        if (gpsSize > 0) {
+            stream = new ByteArrayInputStream(guidePosts.get(), guidePosts.getOffset(), guidePosts.getLength());
+            input = new DataInputStream(stream);
+            if (gps.getMaxLength() > 0) {
+                decoder = new PrefixByteDecoder(gps.getMaxLength());
+            } else {
+                decoder = new PrefixByteDecoder();
+            }
```

Comment:
Instead of this if statement everywhere, just do this (and ensure that PrefixByteDecoder does the right thing if the maxLength is zero):

    decoder = new PrefixByteDecoder(gps.getMaxLength());